java-checks/src/main/java/org/sonar/java/checks/StringIndexofRangesCheck.java

```diff
@@ -0,0 +1,187 @@
+    }
+
+    if (beginIdxConst.isPresent() && endIdxConst.isPresent()
+      && ((int) beginIdxConst.get()) > ((int) endIdxConst.get())) {
+      reportBeginLargerThanEnd(methodInvocation, beginIdxExpr, endIdxExpr);
+    }
+
+    if (receiverConst.isPresent() && beginIdxConst.isPresent()
+      && receiverConst.get() instanceof String s
+      && ((int) beginIdxConst.get()) >= s.length()) {
+      reportIssue(beginIdxExpr, "Begin index should be smaller than string length.",
+        callAsSingletonSecondaryLocation(methodInvocation), null);
+    }
+
+    if (receiverConst.isPresent() && endIdxConst.isPresent()
+      && receiverConst.get() instanceof String s
+      && ((int) endIdxConst.get()) > s.length()) {
+      reportIssue(endIdxExpr, "End index should not be larger than string length.",
+        callAsSingletonSecondaryLocation(methodInvocation), null
+      );
+    }
+
+  }
+
+  private void checkBoundsThatDependOnLength(MethodInvocationTree methodInvocation, String lengthReceiverVarName) {
+    var beginIdxExpr = methodInvocation.arguments().get(1);
+    var endIdxExpr = methodInvocation.arguments().get(2);
+    var beginIdxDelta = lengthDelta(beginIdxExpr, lengthReceiverVarName);
+    var endIdxDelta = lengthDelta(endIdxExpr, lengthReceiverVarName);
+    if (beginIdxDelta.isPresent() && beginIdxDelta.get() >= 0) {
+      reportIssue(beginIdxExpr, "Begin index should be smaller than the length of the string.",
+        callAsSingletonSecondaryLocation(methodInvocation), null);
+    }
+    if (endIdxDelta.isPresent() && endIdxDelta.get() > 0) {
+      reportIssue(endIdxExpr, "End index should be at most the length of the string.",
+        callAsSingletonSecondaryLocation(methodInvocation), null);
+    }
+    if (beginIdxDelta.isPresent() && endIdxDelta.isPresent() && beginIdxDelta.get() > endIdxDelta.get()) {
+      reportBeginLargerThanEnd(methodInvocation, beginIdxExpr, endIdxExpr);
+    }
+  }
+
+  private static List<JavaFileScannerContext.Location> callAsSingletonSecondaryLocation(MethodInvocationTree methodInvocation) {
+    return List.of(new JavaFileScannerContext.Location("Call", methodInvocation));
+  }
+
+  private void reportBeginLargerThanEnd(MethodInvocationTree methodInvocation, ExpressionTree beginIdxExpr, ExpressionTree endIdxExpr) {
+    reportIssue(endIdxExpr, "Begin index should not be larger than endIndex.",
+      List.of(
+        new JavaFileScannerContext.Location("Call", methodInvocation),
+        new JavaFileScannerContext.Location("Begin index", beginIdxExpr)
+      ), null
+    );
+  }
+
+  /**
+   * @param expr    the expression
+   * @param varName name of the variable
+   * @return an Optional containing the difference delta such that the value of expr is var.length() + delta (delta may be negative),
+   * or an empty Optional if the expression is too complex for delta to be computed
+   */
+  private static Optional<Integer> lengthDelta(ExpressionTree expr, String varName) {
```

Comment:
I think we can simplify this method a bit by leveraging what is already done in `SyntacticEquivalence.areEquivalentIncludingSameVariables` to understand if parts of what we are summing up have the same value